agera/src/main/java/com/google/android/agera/FunctionCompiler.java

```diff
@@ -197,52 +190,99 @@ public Function thenSort(@NonNull final Comparator comparator) {
     @SuppressWarnings("unchecked")
     public Object apply(@NonNull final Object input) {
       Object item = input;
-      for (final Function function : functions) {
-        item = function.apply(item);
-      }
-      return item;
-    }
-  }
-
-  private static final class FilterFunction<T> implements Function<List<T>, List<T>> {
-    @NonNull
-    private final Predicate filter;
+      boolean compilerList = false;
+      int i = 0;
+      final int length = functions.length;
+      while (i < length) {
+        final int type = (int) functions[i++];
+        final Object command = functions[i++];
+        switch (type) {
+          case LIMIT: {
+            final List list = (List) item;
+            final int limit = (int) command;
+            if (limit <= 0) {
+              item = emptyList();
+            } else if (list.size() < limit) {
+              item = input;
+            } else if (compilerList) {
+              while (list.size() > limit) {
+                list.remove(list.size() - 1);
+              }
+            } else {
+              final ArrayList newList = new ArrayList(limit);
+              for (int copied = 0; copied < limit; copied++) {
+                newList.add(list.get(copied));
+              }
+              item = newList;
+              compilerList = true;
+            }
+          }
+          break;
+          case FILTER: {
+            final List list = (List) item;
+            if (list.isEmpty()) {
+              item = emptyList();
+            } else if (compilerList) {
+              final Predicate predicate = (Predicate) command;
```

Comment:
hm, O(N^2) in what way? Isn't that just assuming that remove is implemented by indexing over values, either in java or native space, with no access to a memcpy? Actually not sure how much an arraycopy (that's really used here, and this implies just shuffling memory at assembler level to me) would cost. It's what made me not do inline editing here in the first PS.
I'll let you and @maxtroy fight this one out ;) (even if this particular piece of code is from me :P)